### ISSUE TITLE
Add an example for custom executor

### ISSUE THREAD

**Issue by pegasas** [apache/airflow#33146]
### What do you see as an issue?

Hi, Community,

I've been trying to add my custom executor recently, searching in the network, and only found there is no 'How-To-Guides' for this feature.

https://airflow.apache.org/docs/apache-airflow/stable/core-concepts/executor/index.html#executor-types
![image](https://github.com/apache/airflow/assets/13224827/6d7b1053-7d1a-4cdf-b153-bcf48e97af6d)

I guess it should be similar to https://airflow.apache.org/docs/apache-airflow/stable/howto/custom-operator.html

But I would like here to add an page link to howto, and add a related page here.

What's your suggestion?

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by Lee-W** [comment]
Yep, I think it would be great to have more detailed documentation on it. 🙂

**Comment by potiuk** [comment]
Yeah @o-nikolas  - I think that would be a good addition as part of the AWS executor to write a short "howto" based on the learnings . That would be truly completing AIP-51 finally.

**Comment by raphaelauv** [comment]
```python
from threading import Thread
from typing import Optional

from airflow.executors.local_executor import LocalExecutor
from airflow.exceptions import AirflowException
from airflow.jobs.job import Job, run_job
from airflow.jobs.triggerer_job_runner import TriggererJobRunner


class LocalAsyncExecutor(LocalExecutor):

    def __init__(self, **kwargs):
        super().__init__(**kwargs)

        self.thread_triggerer_job: Optional[Thread] = None
        self.triggerer_job_runner: Optional[TriggererJobRunner] = None

    def start(self) -> None:
        super().start()
        self.triggerer_job_runner = TriggererJobRunner(Job())
        self.thread_triggerer_job = Thread(
            target=run_job,
            kwargs={"job": self.triggerer_job_runner.job, "execute_callable": self.triggerer_job_runner._execute})
        self.thread_triggerer_job.start()

    def sync(self) -> None:
        super().sync()

        if self.triggerer_job_runner.trigger_runner.stop:
            raise AirflowException("triggerer process no more running")

    def end(self) -> None:
        self.triggerer_job_runner.on_kill()
        self.thread_triggerer_job.join()
        super().end()
 ```

**Comment by pegasas** [comment]
Hi, @raphaelauv , 

Shall the community not recommend directly inheriting from [BaseExecutor](https://github.com/apache/airflow/blob/main/airflow/executors/base_executor.py)?

Are there any potential risks in it?

**Comment by raphaelauv** [comment]
`Are there any potential risks in it?`

this is a power user functionality andI would not be surprise if AWS do 

```python
class AwsEcsExecutor(CeleryExecutor):
```

for https://github.com/apache/airflow/issues/32943

**Comment by o-nikolas** [comment]
> Yeah @o-nikolas - I think that would be a good addition as part of the AWS executor to write a short "howto" based on the learnings . That would be truly completing AIP-51 finally.

@potiuk, yupp totally agree! It's been a task I had on the project board from the very beginning :)  #27934

Hoping to get to it soon now that the CLI stuff is close to resolution. I've just been distracted by some triggerer changes I'm trying to get merged in time for 2.7.0

> Shall the community not recommend directly inheriting from [BaseExecutor](https://github.com/apache/airflow/blob/main/airflow/executors/base_executor.py)?

We do indeed recommend this (or _will_ do so soon anyways :smile:), the BaseExecutor is the public interface to create your own executor, it's the thing that we'll make a harder promise on backwards compatibility.

> yes this is a power user functionality and I would not be surprise if AWS do

> class AwsEcsExecutor(CeleryExecutor):

We are indeed working on a new ECS executor (which will be a great new example of implementing a third part executor) but we do **not** inherit from the CeleryExecutor. Celery executor is an executor from a different provider (very recently) it is not part of core and does not guarantee as much backwards compat as the BaseExecutor public Interface in core Airflow. Nor is it a very good starting point since it's quite a complicated executor and comes with lots of machinery that you don't need or would complicate your executor.

**Comment by pegasas** [comment]
> We are indeed working on a new ECS executor (which will be a great new example of implementing a third part executor) but we do not inherit from the CeleryExecutor. Celery executor is an executor from a different provider (very recently) it is not part of core and does not guarantee as much backwards compat as the BaseExecutor public Interface in core Airflow. Nor is it a very good starting point since it's quite a complicated executor and comes with lots of machinery that you don't need or would complicate your executor.

It would be awesome if the guideline would start from BaseExecutor! Thanks @o-nikolas 

> @potiuk, yupp totally agree! It's been a task I had on the project board from the very beginning :) https://github.com/apache/airflow/issues/27934

So this should be tracked by https://github.com/apache/airflow/issues/27934.
